### ISSUE TITLE
`clone_on_copy` gives wrong suggestion, causes error

### ISSUE THREAD

**Issue by teofr** [rust-lang/rust-clippy#15610]
### Summary

`clone_on_copy` checks if a type implements copy disregarding the lifetimes, causing it to fail in certain cases 

---

From what I could see this problem is not only related to `clone_on_copy`, but to every lint that uses the `is_copy` function, since it doesn't take into account regions. It also seems to be related to #15577 

I'm not sure if there's any reason to not consider lifetimes when checking trait implementations, but if it's a decision across clippy it may affect more lints.

### Lint Name

clone_on_copy

### Reproducer

I tried this code:

```rust
#![no_main]

struct Weird<'a>(&'a i32);

impl Clone for Weird<'_> {
    fn clone(&self) -> Self {
        Weird(self.0)
    }
}

impl Copy for Weird<'static> {}

impl Weird<'_> {
    fn foo(&self) -> Self {
        self.clone()
    }
}
```

I saw this happen:

```
warning: using `clone` on type `Weird<'_>` which implements the `Copy` trait
  --> test3.rs:15:9
   |
15 |         self.clone()
   |         ^^^^^^^^^^^^ help: try dereferencing it: `*self`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#clone_on_copy
   = note: `#[warn(clippy::clone_on_copy)]` on by default
```

After applying the fixes compilation fails:

```
error: lifetime may not live long enough
  --> test3.rs:15:9
   |
14 |     fn foo(&self) -> Self {
   |            ----- has type `&Weird<'1>`
15 |         *self
   |         ^^^^^ copying this value requires that `'1` must outlive `'static`

error: aborting due to 1 previous error
```

I expected to see this happen:

*No lint*

### Version

```text
rustc 1.91.0-nightly (6ba0ce409 2025-08-21)
binary: rustc
commit-hash: 6ba0ce40941eee1ca02e9ba49c791ada5158747a
commit-date: 2025-08-21
host: aarch64-apple-darwin
release: 1.91.0-nightly
LLVM version: 21.1.0
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by Jarcho** [comment]
This is not at all straightforward to handle. Whether a trait is implemented for a type is independent of whether lifetime constraints are met. Lints currently don't have access to borrowck results which is what we would need to start working on this.

**Comment by teofr** [comment]
I was approaching a similar conclusion after trying to work on this for a bit. Started a discussion [on Zulip](https://rust-lang.zulipchat.com/#narrow/channel/257328-clippy/topic/Trait.20resolution.20ignores.20lifetimes.20constraints/with/537921246).

I'll create a tracking issue for the lifetimes issues (if that's ok with people) and probably unassign this from me for the time being.

**Comment by teofr** [comment]
I don't think I have time to tackle this now. @Jarcho would you mind explaining a bit how checking whether a trait is implemented for a type is independent of lifetime constraints? I'm quite new to the rust world, so apologies if that's obvious.

@rustbot release-assignment

**Comment by Jarcho** [comment]
Forgot to respond to this. Basically the type system doesn't take specific lifetimes into account at all. Both `impl<'a> Copy for S<'a>` and `impl Copy for S<'static>` are both seen by type system as effectively `impl Copy for S<'0>` where `'0` is some lifetime. In the case of `impl Copy for S<'static>` there is an additional `'0: 'static` bound which is known to exist, but is only evaluated later by the borrow checker. This effectively means that in both cases `Copy` is implemented for all lifetimes.

**Comment by arunsingh** [comment]
@Jarcho @teofr @Is this completed or yet to be concluded, Whats the status? I would like to take this issue.

**Comment by teofr** [comment]
I haven't worked on it, unaware if other people have. I'd say if it's still reproducible and unassigned then you could go for it.
